src/wire/dhcp.rs

```diff
@@ -0,0 +1,629 @@
+// See https://tools.ietf.org/html/rfc2131 for the DHCP specification.
+
+use byteorder::{ByteOrder, NetworkEndian};
+
+use {Error, Result};
+use super::{EthernetAddress, Ipv4Address};
+use super::arp::Hardware;
+
+const DHCP_MAGIC_NUMBER: u32 = 0x63825363;
+
+enum_with_unknown! {
+    /// The possible opcodes of a DHCP packet.
+    pub enum OpCode(u8) {
+        Request = 1,
+        Reply = 2,
+    }
+}
+
+enum_with_unknown! {
+    /// The possible message types of a DHCP packet.
+    pub enum MessageType(u8) {
+        Discover = 1,
+        Offer = 2,
+        Request = 3,
+        Decline = 4,
+        Ack = 5,
+        Nak = 6,
+        Release = 7,
+        Inform = 8,
+    }
+}
+
+impl MessageType {
+    fn opcode(&self) -> OpCode {
+        match *self {
+            MessageType::Discover | MessageType::Inform | MessageType::Request |
+                MessageType::Decline | MessageType::Release => OpCode::Request,
+            MessageType::Offer | MessageType::Ack | MessageType::Nak => OpCode::Reply,
+            MessageType::Unknown(_) => OpCode::Unknown(0),
+        }
+    }
+}
+
+/// A representation of a single DHCP option.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum DhcpOption<'a> {
+    EndOfList,
+    Pad,
+    MessageType(MessageType),
+    RequestedIp(Ipv4Address),
+    ServerIdentifier(Ipv4Address),
+    Other { kind: u8, data: &'a [u8] }
+}
+
+impl<'a> DhcpOption<'a> {
+    pub fn parse(buffer: &'a [u8]) -> Result<(&'a [u8], DhcpOption<'a>)> {
+        // See https://tools.ietf.org/html/rfc2132 for all possible DHCP options.
+
+        let (length, option);
+        match *buffer.get(0).ok_or(Error::Truncated)? {
+            field::OPT_END => {
+                length = 1;
+                option = DhcpOption::EndOfList;
+            }
+            field::OPT_PAD => {
+                length = 1;
+                option = DhcpOption::Pad;
+            }
+            kind => {
+                length = *buffer.get(1).ok_or(Error::Truncated)? as usize;
+                let data = buffer.get(2..(2+length)).ok_or(Error::Truncated)?;
+                match (kind, length) {
+                    (field::OPT_END, _) |
+                    (field::OPT_PAD, _) =>
+                        unreachable!(),
+                    (field::OPT_DHCP_MESSAGE_TYPE, 1) => {
+                        option = DhcpOption::MessageType(MessageType::from(data[0]));
+                    },
+                    (field::OPT_REQUESTED_IP, 4) => {
```

Comment:
We also need to verify that the length of `data` is 4 right? In theory `length` could be 4, but `data.len()` could be something else right?